### DISCUSSION TITLE
Tray-icon researches

### DISCUSSION THREAD

**Discussion by rusty-snake** [netblue30/firejail#4053] [Show and tell]
### There are many tray-icon standards:

 - [XEmbed](https://specifications.freedesktop.org/xembed-spec/xembed-spec-latest.html).
   - is legacy
   - uses X 
   - has no wayland support therefore
   - independent on DBus
 -  A Unity/Ubuntu special standard
    - not implemented in Xfce, KDE, ...
    - applications cann't relay on it
 - [StatusNotifierItem (freedesktop)](https://freedesktop.org/wiki/Specifications/StatusNotifierItem/) (`org.freedesktop.StatusNotifierWatcher`)
    - actually there is no implementation by Apps or DE (has xfce one?)
    - was submitted by kde based on the one below
 - KStatusNotifierItem (KDE) (`org.kde.StatusNotifierWatcher`)
    - is used by sway, GNOME [via shell-extension](https://github.com/ubuntu/gnome-shell-extension-appindicator), KDE, ...
    - KDE has maybe some additional additions
    - I tested under GNOME for now
    - this it the one I focus below
 - ...

### Which dbus rules are requiered for KStatusNotifierItem?

- `dbus-user.talk org.kde.StatusNotifierWatcher`
  - This is **unsafe** and allows sandbox escapes
  - This is always required 
- `dbus-user.own <APP-ID>` (or `dbus-user.own <Apps Indicator Service>` but I never saw it)
  - This is required by apps using libayatana (e.g. transmission-gtk)
- `dbus-user.own org.kde.StatusNotifierItem-<PID>-<ID>`
  - This is required by apps strictly following the spec (e.g. Qt based)
  - YES YOU READ IT RIGHT, `<PID>`!! fuck yeah!
  - The pid is part of the name, meaning you need to add `dbus-user.own org.kde.*` (obviously **unsafe**)
  - The pid _inside_ the sandbox is used, this means
    - These names are a bit stable if it is for example the second process in the sandbox which registers a tray-icon
    - Names can conflict and apps likely don't expect a other program with the same pid

**Comment by rusty-snake** [comment]
> The pid is part of the name, meaning you need to add `dbus-user.own org.kde.*` (obviously **unsafe**)

The alternative is to add hundreds of `dbus-user.own org.kde.StatusNotifierItem-<PID>-<ID>` commands. However, this takes endless to start firejal.

```python3
with open("/proc/sys/kernel/pid_max", "r") as pid_max_fd:
    pid_max = int(pid_max_fd.read())

with open("snw.inc", "w") as snw_inc:
    snw_inc.write("dbus-user.call org.kde.StatusNotifierWatcher=org.freedesktop.DBus.Introspectable.*\n")
    snw_inc.write("dbus-user.call org.kde.StatusNotifierWatcher=org.freedesktop.DBus.Properties.*\n")
    snw_inc.write("dbus-user.call org.kde.StatusNotifierWatcher=org.kde.StatusNotifierWatcher.*@/StatusNotifierWatcher\n")
    for pid in range(pid_max):
        for i in range(9):
            snw_inc.write(f"dbus-user.own org.kde.StatusNotifierItem-{pid}-{i}\n")
```

**Comment by crocket** [comment]
Have you found a way to sandbox applications that use StatusNotifierItem?

**Reply by rusty-snake** [reply]
You can sandbox them. But you need to create very broad D-Bus permissions. That's why I created #4510, one some indicator programs don't work without tray icon, everything else works.

**Reply by crocket** [reply]
Did you have to give up PID namespace for softwares that use StatusNotifierItem?

**Reply by rusty-snake** [reply]
Why do you think that?

**Reply by crocket** [reply]
How did you avoid giving up PID namespace or slowing down firejail?

**Reply by rusty-snake** [reply]
Why do you want to giving up PID namespace?

**Reply by crocket** [reply]
I don't want to give up PID namespace or slow down firejail. I wanted to know how you avoided giving up PID namespace or slowing down firejail.

**Reply by rusty-snake** [reply]
Why do you think it is necessary to giving up PID namespace?

**Reply by crocket** [reply]
You said it was unsandboxable. I thought it was due to PID clashes between PID namespaces on dbus.
You also said resolving PID clashes on dbus slowed down firejail.

How did you succeed in sandboxing applications that use StatusNotifierItem?

**Reply by rusty-snake** [reply]
> You said it was unsandboxable.

You can not sandbox this API. However, you can sandbox a program using the API.

> I thought it was due to PID clashes between PID namespaces on dbus.

As long as they don't happen ...
Also IDK what will actually happen on clashes.

> You also said resolving PID clashes on dbus slowed down firejail.

Where? If you refer to https://github.com/netblue30/firejail/discussions/4053#discussioncomment-435987, this is to avoid some of the unsafeties of the API. But it still have possible clashes.

> How did you succeed in sandboxing applications that use StatusNotifierItem?

`grep Status /etc/firejail/*`

**Reply by crocket** [reply]
So, the problem still exists.

**Reply by rusty-snake** [reply]
Yes, but I never saw a clash on my system (but I only have 1 or 2 programs using the sys-tray).

If you have problems (do you have one?) you can `firejail --profile=abcd sh -c "( :; ( :; ( abcd ) ) )"` as workaround.

**Reply by crocket** [reply]
Even if PID is the same, ID can be different. But, clashes can still happen.

> StatusNotifierItem
>
> Each application can register an arbitrary number of Status Notifier Items by registering on the session bus the service org.freedesktop.StatusNotifierItem-PID-ID, where PID is the process id of the application and ID is an arbitrary numeric unique identifier between different instances registered by the same application.

Perhaps, you should talk with freedesktop guys.

**Reply by rusty-snake** [reply]
> **actually there is no implementation by Apps or DE**  <sub>(_see OP_)</sub>

**Reply by crocket** [reply]
The KDE version is similar, and it is easy to change specification when there is no implementation.

**Reply by rusty-snake** [reply]
IMHO there should be something like a `org.freedesktop.portal.StatusNotifierItem`.

**Reply by crocket** [reply]
I encourage you to talk about this with freedesktop guys. I don't know much about this.

**Comment by michelesr** [comment]
@rusty-snake can you provide detail on how `dbus-user.talk org.kde.StatusNotifierWatcher` allows sandbox escaping?

**Reply by rusty-snake** [reply]
The most implementations do not care about sandboxing and expose all objects/interfaces on all names. That means it effectively allows other names like `org.gnome.Shell` too.

```
Denied: dbus-send --print-reply --dest=org.gnome.Shell /org/gnome/Shell org.gnome.Shell.Eval string:'"malicious payload"'
Allowed: dbus-send --print-reply --dest=org.kde.StatusNotifierWatcher /org/gnome/Shell org.gnome.Shell.Eval string:'"malicious payload"'
```

**Reply by michelesr** [reply]
is that true also for other things such as `org.freedesktop.Notifications`? doesn't that make the whole dbus filtering useless, making dbus disabling the only viable option?

**Reply by rusty-snake** [reply]
> org.freedesktop.Notifications

https://github.com/netblue30/firejail/wiki/Restrict-DBus

> doesn't that make the whole dbus filtering useless

- No, you can still use filter to allow IPC (e.g. `dbus-user.own org.mozilla.firefox.*`)
- There are safe names. Mostly DE provided names are affected.

> making dbus disabling the only viable option?

(unlike flatpak,) firejail allows you to specify `call` and `broadcast` policies to filter the objects,interfaces,methods on a name that should be allowed. However they are long, complex, ugly and hard to maintain, that's why they are not used by upstream profiles ATM.

**Comment by opqriu** [comment]
os: debian 12 gnome 43

I started seeing the steam tray icon properly by doing the following.

-.

In the path "/var/lib/dpkg/info", look for a "list" file with the app name.

Found the file "steam-launcher.list"! 

Open it with an editor and search for "png", and you'll find the suspect.

"/usr/share/pixmaps/steam_tray_mono.png"

Put the file "steam_tray_mono.png" in the path /home/firejail/Steam/.local/share/icons. 

How I run steam : 
~$ firejail --private=$HOME/firejail/Steam steam -forcedesktopscaling 2.0 -silent

-.

The tray icon shows the steam image normally. (Previously, when running with firejail --private, the tray icon image was not displayed correctly).

But... 🤔🤔🤔🤔🤔
Tray icon images are very visible even after deleting the /home/firejail/Steam/.local/share/icons folder.

I'm not sure how or why this worked for me. Anyway, that's good
It worked, but I don't know why it worked. Anyway, that's good

**Comment by lucasmz-dev** [comment]
Should this be added to https://github.com/netblue30/firejail/wiki/Restrict-DBus? It seems a bit out of date and it'd be nice to have it updated as well for anything else known by now to be honest

**Reply by rusty-snake** [reply]
Has anything changed?

**Reply by lucasmz-dev** [reply]
I'm not sure. I think not? Pretty much using KDE and tray icons I've only seen work with either X11 or `org.kde.StatusNotifierWatcher`. I'm not very knowledgeable there.
I do see a lot of apps using tray icons, usually those not just made for GNOME, and that feels like something worth noting like `org.freedesktop.Notifications`.

Note: this was as in `org.kde.StatusNotifierWatcher`

ATM I'm migrating from KDE to GNOME just because the sandboxing situation doesn't seem great outside GNOME 🥴

**Comment by ilikenwf** [comment]
@rusty-snake Would it be possible to allow either passing args to the application being executed, or add an option somewhere (or to the current tray allowance) that lets us pass --start-in-tray? Most chat applications are crappy electron applications that require this not to steal focus.

**Reply by rusty-snake** [reply]
If the application supports this. Otherwise not.
